internal/infrastructure/kubernetes/infra_resource.go

```diff
@@ -69,6 +69,11 @@ func (i *Infra) createOrUpdateDeployment(ctx context.Context, r ResourceRender)
 		return err
 	}
 
+	// skip deployment creation if it is not needed
+	if deployment == nil {
```

Comment:
Not very intuitive that `createOrUpdateDeployment` may actually delete the deployment. I guess that this is used for a "cleanup" scenario, but maybe we should be more explicit about how DS/DP is deleted when switching between modes in the top-level reconciling flow calling these methods

Reply:
I can change name of function `createOrUpdateSomething` to `ensureSomething`. Then we will avoid logical errors.

Reply:
And! `CreateOrUpdateHPA` [already has](https://github.com/envoyproxy/gateway/blob/main/internal/infrastructure/kubernetes/infra_resource.go#L66) delete logic)